source/sigv4.c

```diff
@@ -793,6 +854,230 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+                }
+                else
+                {
+                    *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ];
+                }
+
+                pBufLoc++;
+            }
+
+            *pBufLoc = ';';
+            pBufLoc++;
+            buffRemaining -= ( keyLen + 1 );
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = buffRemaining;
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, keyLen = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( ( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData ) != NULL )
+            sigV4Status = writeSignedHeaderToString( noOfHeaders, flags, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            canonicalRequest->pBufCur--;
+            *canonicalRequest->pBufCur = '\n';
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t writeCanonicalHeaderToString( size_t headerIndex,
+                                                       CanonicalContext_t * canonicalRequest )
+    {
+        size_t buffRemaining, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0;
+        char * pBufLoc;
+        const char * value;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+        valLen = canonicalRequest->pHeadersLoc[ headerNum ].value.dataLen;
+
+        if( keyLen + valLen + 2 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
+                pBufLoc++;
+            }
+
+            *pBufLoc = ':';
+            pBufLoc++;
+            value = canonicalRequest->pHeadersLoc[ headerNum ].value.pData;
+            trimValueLen = 0;
+
+            for( i = 0; i < valLen; i++ )
+            {
+                if( !( isspace( value[ i ] ) && ( i + 1 <= valLen ) && ( ( i + 1 == valLen ) || isspace( value[ i + 1 ] ) || ( trimValueLen == 0 ) ) ) )
```

Comment:
> To create the canonical headers list, convert all header names to lowercase and remove leading spaces and trailing spaces. Convert sequential spaces in the header value to a single space.

I think this piece of code just removes all spaces from the string and needs more testing

Reply:
Please update this expression into a more readable form by creating private functions that allow the expression to look like:
`if ( ! isLeadingSpace( value, i ) && !  isTrailingSpace( value, i ) && ! isFirsInSequentialSpaces( value, i ) )`

Reply:
> I think this piece of code just removes all spaces from the string and needs more testing

The condition inside the `if` is basically:
```
!( isCurrentCharSpace && ( true ) && ( ( isLastChar ) || isNextCharSpace || ( isLeadingCharacter ) ) )
```
which further boils down to:
```
!( isCurrentCharSpace && ( isLastChar || isNextCharSpace || isLeadingCharacter ) )
```
which looks like it would correctly trim spaces, it's just confusing to think about and could probably be replaced with a function to be more readable

Reply:
Copying my suggestion from slack:
```C
static bool isTrimmableSpace( const char * value, size_t index, size_t valLen, size_t trimmedLength ) {
    bool ret = false;

    /* Only trim spaces. */
    if( isspace( value[ index ] ) )
    {
        /* The last character is a trailing space. */
        if( ( index + 1 ) == valLen )
        {
            ret = true;
        }
        /* Trim if the next character is also a space. */
        else if( isspace( value[ index + 1 ] ) )
        {
            ret = true;
        }
        /* It is a leading space if no characters have been written yet. */
        else if( trimmedLength == 0U )
        {
            ret = true;
        }
    }
    return ret;
}
```